### DISCUSSION TITLE
Traceback caused by A Missing Required Env reveals other Envs to stdout

### DISCUSSION THREAD

**Discussion by tea-418** [fastapi/fastapi#12922] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic_settings import BaseSettings
import os
from fastapi import FastAPI

os.environ["API_SECRET"] = 'SuperSecretSecretIHopeNoOneReadsThis'

class Settings(BaseSettings):
    api_url: str = os.getenv('API_URL', 'https://super-cool-api.com/v1/')
    api_user: str = os.getenv('API_USER', 'john-doe')
    api_secret: str = os.getenv('API_SECRET')
    different_secret: str = os.getenv('DIFFERENT_SECRET')

settings = Settings()

app = FastAPI()
```

### Description

* Execute with _fastapi run_.
* It Fails because of a ValidationError of Pydantic - as it should - because "different_secret" is empty.
* The Output contains secrets I wish not to reveal in logging.

### Operating System

macOS

### FastAPI Version

0.115.3

### Pydantic Version

2.9.2

### Python Version

Python 3.12.2

### Additional Context

Hey everybody!
I followed the [Tutorial](https://fastapi.tiangolo.com/advanced/settings/?h=env#settings-and-environment-variables) for using Environment Variables with FastAPI. In Order to use some secrets I need for my backend services.

However whenever a Env is not set that is expected to be set, an output like this will be produced:
![Bildschirmfoto 2024-11-13 um 13 29 34](https://github.com/user-attachments/assets/aa682243-7e53-4480-9e33-adae875d5101)

**My Issue is the following:** The Traceback reveals the locals which unfortunately includes the "api_secret" which I dont want to be logged. 

How can I stop this from happening, I tried setting the type to _SecretStr_ and also not instancing the Pydantic _BaseSetting_ class but instead raise a Error manually but both still caused the locals to be logged to console.

I would be happy disabling the logging of locals all together or also disabling the Traceback altogether but I dont know how I can do that.

I am aware that rich.traceback has a parameter "show_locals" which I assume would help me here but for the life of me I am unable to find a way to modify the logging of FastAPI in a way that makes it possible for me to hide locals.

**Comment by YuriiMotov** [comment, marked as answer]
I think you can overcome this by wrapping `settings = Settings()` by `try..except`:

```
from pydantic import ValidationError

# code omitted ...

try:
    settings = Settings()
except ValidationError as e:
    for error in e.errors(include_input=False):
        logger.error(error)
```

**Reply by tea-418** [reply]
Thank you, that works for me!

However I'm still wondering why the traceback here looks so different from the traceback that occurs whenever an error happens on a route and if the locals might still be logged in a different scenario I cannot foresee. 
Lastly I think this might be a weird behavior because the Tutorial states that this approach should be used for secrets, which in my opinion should never be logged.

However I will close the discussion because my Issue is solved and Im not sure if this justifies further investigation/concern.

Thank you again for resolving my Issue so quickly, I will add the complete working Code below for anyone having the same Issue.

```
import logging

from pydantic_settings import BaseSettings
from pydantic import SecretStr, ValidationError
import os
from fastapi import FastAPI

os.environ["API_SECRET"] = 'SuperSecretSecretIHopeNoOneReadsThis'

class Settings(BaseSettings):
    api_url: str = os.getenv('API_URL', 'https://super-cool-api.com/v1/')
    api_user: str = os.getenv('API_USER', 'john-doe')
    api_secret: str = os.getenv('API_SECRET')
    different_secret: str = os.getenv('DIFFERENT_SECRET')

logger = logging.getLogger('uvicorn.error')

app = FastAPI()

logger.info('API is starting up')
try:
    settings = Settings()
except ValidationError as e:
    for error in e.errors(include_input=False):
        logger.critical(error)
        exit(1)
```
